api/operator/v1alpha1/istiocsr_types.go

```diff
@@ -0,0 +1,246 @@
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="issuerRef is immutable once set"
+	// +required
+	IssuerRef certmanagerv1.ObjectReference `json:"issuerRef,omitempty"`
+}
+
+// IstiodTLSConfig is for configuring istiod certificate specifics.
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.signatureAlgorithm) && !has(self.signatureAlgorithm) || has(oldSelf.signatureAlgorithm) && has(self.signatureAlgorithm)",message="signatureAlgorithm may only be configured during creation"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.privateKeySize) && !has(self.privateKeySize) || has(oldSelf.privateKeySize) && has(self.privateKeySize)",message="privateKeySize may only be configured during creation"
+type IstiodTLSConfig struct {
+	// commonName is the common name to be set in the certificate.cert-manager.io
+	// created for istiod. CommonName will be of the form `istiod.<istio_namespace>.svc`
+	// when not set.
+	// +optional
+	CommonName string `json:"commonName,omitempty"`
+
+	// trustDomain is the istio cluster's trust domain, which will also be used for deriving
+	// spiffe URI.
+	// +required
+	TrustDomain string `json:"trustDomain,omitempty"`
+
+	// certificateDNSNames contains the additional DNS names to be added to the istiod certificate SAN.
+	// +listType=set
+	// +optional
+	CertificateDNSNames []string `json:"certificateDNSNames,omitempty"`
+
+	// certificateDuration is the istio-csr and the istiod certificates validity period.
+	// +kubebuilder:default:="1h"
+	// +optional
+	CertificateDuration *metav1.Duration `json:"certificateDuration,omitempty"`
+
+	// certificateRenewBefore is the ahead time to renew the istio-csr and the istiod certificates
+	// before expiry.
+	// +kubebuilder:default:="30m"
+	// +optional
+	CertificateRenewBefore *metav1.Duration `json:"certificateRenewBefore,omitempty"`
+
+	// privateKeySize is the istio-csr and the istiod certificate's key size. When the SignatureAlgorithm
+	// is RSA, must be >= 2048 and for ECDSA, can only be 256 or 384, corresponding to P-256 and P-384 respectively.
+	// +kubebuilder:default:=2048
+	// +kubebuilder:validation:XValidation:rule="oldSelf == 0 || self == oldSelf",message="privateKeySize is immutable once set"
+	// +optional
+	PrivateKeySize int `json:"privateKeySize,omitempty"`
+
+	// signatureAlgorithm is the signature algorithm to use when generating
+	// private keys. At present only RSA and ECDSA are supported.
+	// +kubebuilder:default:="RSA"
+	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
+	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="signatureAlgorithm is immutable once set"
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration *metav1.Duration `json:"maxCertificateDuration,omitempty"`
```

Comment:
How is `MaxCertificateDuration` different from `CertificateDuration`. This field is not used to set any value of the corresponding certificates.

Reply:
`CertificateDuration` is the validity period of a certificate and `MaxCertificateDuration` is the maximum validity duration that can be requested for a certificate. This is used by `istio-csr` to impose limit on the `CertificateDuration`.
This config is assigned as arg value to `istio-csr` [here](https://github.com/openshift/cert-manager-operator/pull/220/files/5983f1bd8cbe7e061f8c21831c7e9bc10f372075..5741bd302d960355cfade4bdbb11b7136d734528#diff-39fcf0b04eed7608202be95c049d3512f02ad20b5764adfc64ec5c3c6aae24e6R145).